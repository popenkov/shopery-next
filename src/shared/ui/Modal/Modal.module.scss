.Modal {
  position: fixed;
  inset: 0;
  z-index: 100;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 0;
  pointer-events: none;
}

.content {
  z-index: var(--modal-content-z-index);
  max-width: 60%;
  padding: 20px;
  border-radius: 12px;
  transform: scale(0.5);
  transition: 0.3s transform;
}

.modalNew {
  .content {
    background: var(--light-bg-redesigned);
  }
}

.modalOld {
  .content {
    background: var(--bg-color);
  }
}

.opened {
  z-index: var(--modal-content-z-index);
  opacity: 1;
  pointer-events: auto;

  .content {
    transform: scale(1);
  }
}

.isClosing {
  .content {
    transform: scale(0.2);
  }
}
