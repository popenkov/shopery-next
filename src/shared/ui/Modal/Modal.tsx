'use client';

import { ReactNode } from 'react';
import cn from 'classnames';

import { useModal } from '@/shared/lib/hooks';

import { Overlay } from '../Overlay';
import { Portal } from '../Portal';

import cls from './Modal.module.scss';

interface ModalProps {
  className?: string;
  children?: ReactNode;
  isOpen?: boolean;
  onClose?: () => void;
  lazy?: boolean;
}

const ANIMATION_DELAY = 300;

export const Modal = (props: ModalProps) => {
  const { className, children, isOpen, onClose, lazy } = props;

  const { close, isClosing, isMounted } = useModal({
    animationDelay: ANIMATION_DELAY,
    onClose,
    isOpen,
  });

  const mods = {
    [cls.opened]: isOpen,
    [cls.isClosing]: isClosing,
  };

  if (lazy && !isMounted && typeof window === 'undefined') {
    return null;
  }

  return (
    <Portal element={document.getElementById('app') ?? document?.body}>
      <div className={cn(cls.Modal, mods)}>
        <Overlay onClick={close} />
        <div className={cn(cls.content, className)}>{children}</div>
      </div>
    </Portal>
  );
};
