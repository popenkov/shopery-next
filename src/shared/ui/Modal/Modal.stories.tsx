import type { Meta, StoryObj } from '@storybook/react';

import { Modal } from './Modal';

const meta = {
    title: 'shared/Modal',
    component: Modal,
    parameters: {
        layout: 'centered',
    },
    tags: ['autodocs'],
} satisfies Meta<typeof Modal>;

export default meta;
type Story = StoryObj<typeof meta>;

export const ModalComponent: Story = {
    args: {
        isOpen: true,
        children:
            'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aliquid commodi consequatur eligendi impedit incidunt necessitatibus possimus quis saepe sunt totam.\n ',
    },
};
