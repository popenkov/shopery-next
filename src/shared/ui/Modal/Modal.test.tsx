import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Modal } from './Modal';

describe('Modal component', () => {
  it('renders correctly', () => {
    const { container } = render(<Modal isOpen={true}>Modal content</Modal>);
    expect(container).toMatchSnapshot();
  });

  it.skip('calls onClose callback when overlay is clicked', () => {
    const onClose = jest.fn();
    const { getByRole } = render(
      <Modal isOpen={true} onClose={onClose}>
        Modal content
      </Modal>,
    );
    const overlay = getByRole('button');
    fireEvent.click(overlay);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('renders children correctly', () => {
    const { getByText } = render(<Modal isOpen={true}>Modal content</Modal>);
    expect(getByText('Modal content')).toBeInTheDocument();
  });

  it.skip('has correct class names when isOpen is true', () => {
    const { container, debug } = render(<Modal isOpen={true}>Modal content</Modal>);
    // debug();
    // expect(container.querySelector('.Modal')).toHaveClass('opened');
    expect(container).toHaveClass('opened');
  });

  it.skip('has correct class names when isClosing is true', () => {
    const { container, debug } = render(
      <Modal isOpen={true} onClose={() => {}}>
        Modal content
      </Modal>,
    );
    // debug();
    const overlay = container.querySelector('.Overlay');
    fireEvent.click(container);
    // expect(container.querySelector('.Modal')).toHaveClass('isClosing');
    expect(container).toHaveClass('isClosing');
  });

  it('renders null when lazy is true and isMounted is false', () => {
    const { container } = render(
      <Modal lazy={true} isOpen={false}>
        Modal content
      </Modal>,
    );
    expect(container).toBeEmptyDOMElement();
  });

  it('renders correctly when lazy is true and isMounted is true', () => {
    const { container } = render(
      <Modal lazy={true} isOpen={true}>
        Modal content
      </Modal>,
    );
    expect(container).toMatchSnapshot();
  });
});
