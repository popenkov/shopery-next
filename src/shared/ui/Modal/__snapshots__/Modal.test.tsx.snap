// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component renders correctly 1`] = `<div />`;

exports[`Modal component renders correctly when lazy is true and isMounted is true 1`] = `<div />`;
